servant-server/src/Servant/Server/Internal.hs

```diff
@@ -116,11 +118,10 @@ instance (KnownSymbol capture, FromText a, HasServer sublayout)
      a -> ServerT sublayout m
 
   route Proxy subserver =
-    DynamicRouter $ \ first ->
-      route (Proxy :: Proxy sublayout)
-            (case captured captureProxy first of
-               Nothing  -> return $ failWith NotFound
-               Just v   -> feedTo subserver v)
+    DynamicRouter $ \ first -> case captured captureProxy first of
+       Nothing  -> LeafRouter (\_ r -> r $ Fail err404)
+       Just v   -> route (Proxy :: Proxy sublayout) (feedTo subserver v)
+
```

Comment:
That seems strange, but is probably something that should rather be fixed in the method combinator, by delaying the method check until after processing all other delayed checks. But I'm pretty positive that delaying captures is a good thing. Static matching is the fastest dispatch method we have, and first checking all static path components before checking any dynamic ones seems like a huge improvement of the new routing scheme that I'm not willing to easily give up again.